tests/raftstore_cases/test_prevote.rs

```diff
@@ -0,0 +1,216 @@
+use std::sync::{atomic::AtomicBool, mpsc, Arc};
+use std::time::Duration;
+
+use super::cluster::{Cluster, Simulator};
+use super::server::new_server_cluster;
+use super::transport_simulate::*;
+use super::util::*;
+use raft::eraftpb::MessageType;
+
+enum FailureType<'a> {
+    Partition(&'a [u64], &'a [u64]),
+    Reboot(&'a [u64]),
+}
+
+fn attach_prevote_notifiers<T: Simulator>(cluster: &Cluster<T>, peer: u64) -> mpsc::Receiver<()> {
+    // Setup a notifier
+    let (tx, rx) = mpsc::channel();
+    let response_notifier = Box::new(MessageTypeNotifier::new(
+        MessageType::MsgRequestPreVoteResponse,
+        tx.clone(),
+        Arc::from(AtomicBool::new(true)),
+    ));
+    let request_notifier = Box::new(MessageTypeNotifier::new(
+        MessageType::MsgRequestPreVote,
+        tx.clone(),
+        Arc::from(AtomicBool::new(true)),
+    ));
+
+    cluster
+        .sim
+        .write()
+        .unwrap()
+        .add_send_filter(peer, response_notifier);
+    cluster
+        .sim
+        .write()
+        .unwrap()
+        .add_send_filter(peer, request_notifier);
+
+    rx
+}
+
+// Validate that prevote is used in elections after partition or reboot of some nodes.
+fn test_prevote<T: Simulator>(
+    cluster: &mut Cluster<T>,
+    failure_type: FailureType,
+    leader_after_failure_id: impl Into<Option<u64>>,
+    detect_during_failure: (u64, bool),
+    detect_during_recovery: (u64, bool),
+) {
+    cluster.cfg.raft_store.prevote = true;
+
+    let leader_id = 1;
+
+    // We must start the cluster before adding send filters, otherwise it panics.
+    cluster.run();
+
+    cluster.must_transfer_leader(1, new_peer(leader_id, 1));
+    cluster.must_put(b"k1", b"v1");
+
+    // Determine how to fail.
+    let rx = attach_prevote_notifiers(cluster, detect_during_failure.0);
+    match failure_type {
+        FailureType::Partition(majority, minority) => {
+            cluster.partition(majority.clone().to_vec(), minority.clone().to_vec());
+        }
+        FailureType::Reboot(peers) => {
+            peers.iter().for_each(|&peer| cluster.stop_node(peer));
+        }
+    };
+
+    // Once we see a response on the wire we know a prevote round is happening.
+    let received = rx.recv_timeout(Duration::from_secs(3));
+    assert_eq!(
+        received.is_ok(),
+        detect_during_failure.1,
+        "Sends a PreVote or PreVoteResponse during failure.",
+    );
+
+    if let Some(leader_id) = leader_after_failure_id.into() {
+        cluster.must_transfer_leader(1, new_peer(leader_id, 1));
+    }
+
+    // Prepare to listen.
+    let rx = attach_prevote_notifiers(cluster, detect_during_recovery.0);
+
+    // Let the cluster recover.
+    match failure_type {
+        FailureType::Partition(_, _) => {
+            cluster.clear_send_filters();
```

Comment:
Does it remove the filters that attached in L85?

https://github.com/pingcap/tikv/pull/3133/files#diff-83b3a9ffc11374ee9efdaea3710b058bR85

Reply:
Oh, you're right.

Reply:
I've reflected this.